Redis adapter

byroot opened the pull request:
Semian integration for redis-rb


@Sirupsen @fw42 for review please (can totally wait tomorrow)

byroot commented on lib/semian/adapter.rb:
```diff
@@ -0,0 +1,34 @@
+      raise NotImplementedError.new("Semian adapters must implement a `semian_options` method")
+    end
+  end
+end
```
I extracted the common patterns here. Maybe that's too much, you guys tell me.

Sirupsen replied:
I think that makes sense.

byroot commented on lib/semian/adapter.rb:
```diff
@@ -0,0 +1,34 @@
+    private
+
+    def acquire_semian_resource(scope:)
+      return yield if @resource_acquired # Prevent a single adapter from acquiring the resource twice by nesting calls
```
Redis basically does:

```ruby
def get
  ensure_connected do
    actuall_get
  end
end
```

Which ended up doing a 2 nested acquires. I believe we better prevent that.

Sirupsen replied:
Nice catch.

Sirupsen replied:
However what if you do something like:

```ruby
@redis1.pipelined do
  @redis1.set "omg", "hi"
  @redis1.set "lol", "hi"
  @redis2.set "omg", "sup"
end
```

Does `pipeline` actually end up acquiring a connection as well? I'm not completely familiar with how the client handles this situation.

Sirupsen replied:
Or even if you could every do something like:

```ruby
@redis1.set "omg", -> { @redis2.set("i-hope", "nobody does this") }
```

byroot replied:
> Does pipeline actually end up acquiring a connection as well?

Pipelined is not a subclass of `Redis::Client`. Also the pipepline are first evaluated and then the client send the commands to the server, e.g:

```ruby
def pipelined
  yield commands_collector
  ensure_connected do
    perform commands_collector.commands
  end
end
```
(This is pseudo code but the idea is here)

> Or even if you could every do something like:

That would be fine too since the flag is an instance variable of the adapter, in this case each redis has his own flag.

fw42 replied:
do we really need the comment here? Do we have a test that captures this possible bug?

Sirupsen replied:
Why not have a comment here @fw42 ? It was helpful for me during review.

fw42 replied:
because comments are often considered a code smell and easily get out of sync with the code. but if you think that it really improves the code here, then ok... I would prefer to have a helper method with a more descriptive name rather than a comment.. something like
```ruby
return yield if same_adapter_already_acquired_resource?
```

fw42 replied:
back to my other question: are any tests going to fail if this line is removed?

Sirupsen replied:
In this case I think you're right :+1: in the other case I think it's still very helpful

byroot replied:
> back to my other question: are any tests going to fail if this line is removed?

Yes all of the `redis_test.rb` suite will fail.

byroot commented:
Also since we have a couple redis clients not using  semian yet, we might have to merge https://github.com/Shopify/semian/pull/25 too.

Sirupsen commented on lib/semian/redis.rb:
```diff
@@ -0,0 +1,58 @@
+    end
+
+    def io(&block)
+      acquire_semian_resource(scope: :io) { raw_io(&block) }
```
Isn't the scope more something like `command` or `query`? A little more human readable when this is used primarily for monitoring.

Sirupsen replied:
We should keep this consistent with MySQL unless it absolutely doesn't make sense.

byroot replied:
Yeah, you're right. `connection` and `query` are probably the 2 scopes most adapters will have. I'll update Mysql2 accordingly.

Sirupsen commented on lib/semian/redis.rb:
```diff
@@ -0,0 +1,58 @@
+      @semian_identifier ||= begin
+        opts = options[:semian] || options['semian'.freeze] || {}
+        name = opts[:name] || opts['name'.freeze]
+        name ||= "#{location}/#{db}"
```
Where does `location` and `db` come from? What are they usually?

byroot replied:
They are implemented by `Redis::Client` https://github.com/redis/redis-rb/blob/be4326e7f2d399069831ba05ef53ba51c9a76c7e/lib/redis/client.rb#L107-L109

`location = "#{host}:#{port}"`. They both have proper defaults.

Sirupsen commented on test/redis_test.rb:
```diff
@@ -0,0 +1,180 @@
+    threads << thread
+    thread.join(0.1)
+    thread
+  end
```
Perhaps it makes sense to extract those to the test helper at this point?

byroot replied:
Indeed, I just copypasted mysql test suite like a lazy bastard. I'll factorize.

Sirupsen commented on test/redis_test.rb:
```diff
@@ -0,0 +1,180 @@
+    end
+  end
+
+  private
```
Needs a test on the nested acquire (although I'm still not sure what that'll look like).

byroot replied:
It's already tested since all redis commands end up doing a nested acquire.

fw42 commented on lib/semian/redis.rb:
```diff
@@ -0,0 +1,58 @@
+
+module Semian
+  module Redis
+    include Adapter
```
Can we ```include Semian::Adapter``` for clarity?

fw42 commented on lib/semian/mysql2.rb:
```diff
@@ -19,6 +13,11 @@ def to_s
+    include Adapter
+
+    ResourceOccupiedError = ::Mysql2::ResourceOccupiedError
+    CircuitOpenError = ::Mysql2::CircuitOpenError
```
oO

fw42 replied:
why do we need this?

byroot replied:
So that `Semian::Adapter` can raise them.

fw42 replied:
seems kinda hacky, but ok

fw42 commented on lib/semian/redis.rb:
```diff
@@ -0,0 +1,58 @@
+    ResourceOccupiedError = ::Redis::ResourceOccupiedError
+    CircuitOpenError = ::Redis::CircuitOpenError
+
+    # The naked methods are exposed as `raw_query` and `raw_connect` for instrumentation purpose
```
comment really necessary?

Sirupsen replied:
I asked for this comment for the MySQL adapter. I think this comment is very helpful.

fw42 commented on test/fixtures/toxiproxy.json:
```diff
@@ -3,5 +3,10 @@
     "name": "semian_test_mysql",
     "upstream": "localhost:3306",
     "listen": "localhost:13306"
+  },
```
hm this file isn't really a fixture, is it? should this be in a different directory?

fw42 commented on test/redis_test.rb:
```diff
@@ -0,0 +1,180 @@
+
+    connect_to_redis!
+
+    assert notified, 'No notification have been emitted'
```
s/have/has/ or s/notification/notifications/

fw42 commented on test/redis_test.rb:
```diff
@@ -0,0 +1,180 @@
+
+    client.get('foo')
+
+    assert notified, 'No notification have been emitted'
```
same

byroot commented:
I just:

  - Refactored the double acquisition handling, with a meaningful method name to get rid of the comment
  - Refactored the test helpers
  - Move the toxiproxy config outside of the fixtures direcotry
  - `include Semian::Adapter` for clarity
  - Normalized scope names with `query` and `connection`

Sirupsen commented on test/test_helper.rb:
```diff
@@ -1,10 +1,17 @@
+
 Semian.logger = Logger.new(nil)
-Toxiproxy.populate(File.expand_path('../fixtures/toxiproxy.json', __FILE__))
+Toxiproxy.populate(File.expand_path('../helpers/toxiproxy.json', __FILE__))
```
Off-side comment: I think `Toxiproxy.populate` should take a Hash, I may have gotten a little aggro in forcing people to store it in a JSON file. I still think that should be done in most use cases, but what do you think about making it a Hash? Would make the introduction simpler too.

byroot replied:
I think the JSON file is handy for project that have many proxies, but yes in that case a literal hash would be simpler.

Sirupsen replied:
Definitely. But at that point people can abstract it themselves easily.

Sirupsen commented:
LGTM

Sirupsen commented:
It'll take some extra work in Shopify because it now means we have to protect all Redises, but that needed to be done anyway.
